### PR TITLE
Extend Ivy Functional API : Linear Algebra functions: matrix_exp

### LINKED ISSUE #7748: matrix_exp

### PR DESCRIPTION
**Pull request by anurag629**
closes #7748

### CODE REVIEW

**Comment by anurag629** [comment]
#7748

**Comment by anurag629** [comment]
> Can you post a link to the issue this pull request is supposed to deal with? Thanks.

@Daniel4078  Done!

File: ivy_tests/test_ivy/test_functional/test_experimental/test_core/test_linalg.py [hunk 1]
```diff
@@ -217,3 +217,40 @@ def test_kron(
...
+    with_out,
+    num_positional_args,
+    native_array,
+    container_flags,
+    instance_method,
+    backend_fw,
+    fn_name,
+):
+    dtype, x = dtype_x
+    helpers.test_function(
```
Comment (Daniel4078) [inline ivy_tests/test_ivy/test_functional/test_experimental/test_core/test_linalg.py:245]: ivy_tests/test_ivy/test_functional/test_experimental/test_core/test_linalg.py:245: in test_matrix_exp
    helpers.test_function(
E   TypeError: test_function() missing 1 required keyword-only argument: 'ground_truth_backend'

**Review by Daniel4078 — CHANGES_REQUESTED** [review]
I found an error that make all of your tests fail, check the comment below.

File: ivy_tests/test_ivy/test_functional/test_experimental/test_core/test_linalg.py [hunk 2]
```diff
@@ -217,3 +217,42 @@ def test_kron(
         a=x[0],
         b=x[1],
     )
+
+
+# matrix_exp
+@handle_test(
+    fn_tree="functional.experimental.matrix_exp",
+    dtype_x=helpers.dtype_and_values(
```
Comment (Daniel4078) [inline ivy_tests/test_ivy/test_functional/test_experimental/test_core/test_linalg.py:225]: E   RuntimeError: linalg.matrix_exp: The input tensor must have at least 2 dimensions.

**Review by Daniel4078 — CHANGES_REQUESTED** [review]
Here is another error reguarding your input array generated for the test that cause all of your test to fail.

File: ivy_tests/test_ivy/test_functional/test_experimental/test_core/test_linalg.py (hunk 1 above)
Comment (Daniel4078) [reply ivy_tests/test_ivy/test_functional/test_experimental/test_core/test_linalg.py:246]: tensorflow: matrix_exp: Value for attr 'T' of uint8 is not in the list of allowed values: bfloat16, half, float, double, int8, int16, int32, int64

Comment (Daniel4078) [reply ivy_tests/test_ivy/test_functional/test_experimental/test_core/test_linalg.py:246]: similarly: jax: matrix_exp: expected A to be a square matrix

File: ivy_tests/test_ivy/test_functional/test_experimental/test_core/test_linalg.py (hunk 2 above)
Comment (Daniel4078) [reply ivy_tests/test_ivy/test_functional/test_experimental/test_core/test_linalg.py:225]: in addition: E   tensorflow.python.framework.errors_impl.InvalidArgumentError: Value for attr 'T' of uint8 is not in the list of allowed values: bfloat16, half, float, double, int8, int16, int32, int64

Comment (Daniel4078) [reply ivy_tests/test_ivy/test_functional/test_experimental/test_core/test_linalg.py:225]: also, I suspect you uses tensorflow's function in your numpy backend version of the function, which is not correct.

**Comment by Daniel4078** [comment]
I am confused, why your test will call the tensorflow backend instead of numpy when testing for numpy?

File: ivy/functional/backends/numpy/experimental/linear_algebra.py [hunk 3]
```diff
@@ -89,3 +89,12 @@ def kron(
 
 
 kron.support_native_out = False
+
+
+def matrix_exp(
```
Comment (Daniel4078) [inline ivy/functional/backends/numpy/experimental/linear_algebra.py:94]: just a small tip, you should check out how to state the unsupported datatypes for these backend functions or find a way to convert them in the function.

Comment (Daniel4078) [reply ivy/functional/backends/numpy/experimental/linear_algebra.py:94]: and I suppose there are ways to specify what datatype you what to generate for your test as well written in the deep dive document.

**Comment by Daniel4078** [comment]
now we have this error to fix:
E   ivy.exceptions.IvyBackendException: torch: matrix_exp: matrix_exp: Expected a floating point or complex tensor as input. Got Byte
E   Falsifying example: test_matrix_exp(
E       dtype_x=(['uint8'], [array([[0, 0],
E                [0, 0]], dtype=uint8)]),
E       num_positional_args=1,
E       as_variable=[False],
E       native_array=[False],
E       with_out=False,
E       container_flags=[False],
E       instance_method=False,
E       backend_fw=<module 'ivy.functional.backends.torch' from '/ivy/ivy/functional/backends/torch/__init__.py'>,
E       fn_name='matrix_exp',
E       ground_truth_backend='tensorflow',
E   )
E   
E   You can reproduce this example by temporarily adding @reproduce_failure('6.55.0', b'AXicY2BABQAAEAAB') as a decorator on your test case

**Comment by anurag629** [comment]
Hey @Daniel4078, I need help in fixing below error.

E   ivy.exceptions.IvyBackendException: tensorflow: matrix_exp: Value for attr 'T' of uint16 is not in the list of allowed values: bfloat16, half, float, double, complex64, complex128
E   	; NodeDef: {{node Exp}}; Op<name=Exp; signature=x:T -> y:T; attr=T:type,allowed=[DT_BFLOAT16, DT_HALF, DT_FLOAT, DT_DOUBLE, DT_COMPLEX64, DT_COMPLEX128]> [Op:Exp]
E   Falsifying example: test_matrix_exp(
E       dtype_x=(['uint16'], [array([[0, 0],
E                [0, 0]], dtype=uint16)]),
E       num_positional_args=1,
E       as_variable=[False],
E       native_array=[False],
E       with_out=False,
E       container_flags=[False],
E       instance_method=False,
E       backend_fw=<module 'ivy.functional.backends.numpy' from '/ivy/ivy/functional/backends/numpy/__init__.py'>,
E       fn_name='matrix_exp',
E       ground_truth_backend='tensorflow',
E   )
E   
E   You can reproduce this example by temporarily adding @reproduce_failure('6.55.0', b'AXicY2RkYAAiZAAAAEcABA==') as a decorator on your test case

**Comment by Daniel4078** [comment]
recently, there is some big change requaring the tests, please have a look:
Hello @everyone,

Just pushed an update of how the test flags are handled, the inference of with_out keyword argument was removed to reduce unneeded performance overhead. 

Some tests will start to fail, i'm working on fixing most of them now but if you are working on a implementing a test and it starts to fail due to unexpected keyword argument 'out' simply remove the with_out keyword and hardcode the test_value to False as done here: https://github.com/unifyai/ivy/commit/fd7ee2daec482b6a56e06b52b395c2bf2777c1ff

Also, some updates will be made in the next few days to how the test flags are passed around, will make things more cleaner and readable. Hopefully no more test factoring after this 😂 , Thanks!

**Comment by Daniel4078** [comment]
another maybe usefull small new for you about the tests:
Aarsh — today 14:14
Hey @everyone, you'll see a couple of tests failing, even though they are passing. I'm pausing the dashboards for the moment. This should be fixed soon, apologies for the inconvenience !

**Comment by anurag629** [comment]
@Daniel4078 why this error and how to fix it.

    
    E   ivy.exceptions.IvyError: tensorflow: matrix_exp: type object 'ContainerBase' has no attribute 'multi_map_in_static_method'
    E   Falsifying example: test_matrix_exp(
    E       dtype_x=(['float64'],
    E        [array([[-1.00000000e+000, -1.00000000e+000, -1.00000000e+000,
    E                 -1.00000000e+000],
    E                [-1.50000000e+000, -1.50000000e+000, -1.50000000e+000,
    E                 -1.50000000e+000],
    E                [-1.50000000e+000, -1.50000000e+000, -1.00000000e+000,
    E                 -1.00000000e+000],
    E                [-1.00000000e+000, -1.00000000e+000, -1.00000000e+000,
    E                 -1.00000000e+000],
    E                [-1.50000000e+000, -1.50000000e+000, -1.50000000e+000,
    E                 -1.50000000e+000],
    E                [-1.00000000e+000, -2.0000[100](https://github.com/unifyai/ivy/actions/runs/3705644627/jobs/6279825331#step:7:101)0e+000, -2.00001000e+000,
    E                 -2.00001000e+000],
    E                [-1.00000000e+000, -2.00001000e+000, -2.00001000e+000,
    E                 -2.00001000e+000],
    E                [-2.00001000e+000, -1.00000000e+000, -1.00000000e+000,
    E                 -1.00000000e+000],
    E                [-1.00000000e+000, -1.00000000e+000, -1.00000000e+000,
    E                 -1.00000000e+000],
    E                [-1.00000000e+000, -1.00000000e+000, -1.00000000e+000,
    E                 -1.00000000e+000],
    E                [-1.00000000e+000, -1.00000000e+000, -1.00000000e+000,
    E                 -1.00000000e+000],
    E                [-1.00000000e+000, -1.00000000e+000, -1.00000000e+000,
    E                 -1.00000000e+000],
    E                [-1.00000000e+000, -1.00000000e+000, -1.00000000e+000,
    E                 -1.00000000e+000],
    E                [-1.00000000e+000, -1.00000000e+000, -1.00000000e+000,
    E                 -1.00000000e+000],
    E                [-1.00000000e+000, -1.00000000e+000, -2.22044605e-016,
    E                 -1.00000000e+000],
    E                [-1.00000000e+000, -1.00000000e+000, -1.00000000e+000,
    E                 -1.00000000e+000],
    E                [-1.00000000e+000, -1.00000000e+000, -1.00000000e+000,
    E                 -1.00000000e+000],
    E                [-1.00000000e+000, -1.00000000e+000, -1.00000000e+000,
    E                 -1.00000000e+000],
    E                [-1.00000000e+000, -1.00000000e+000, -1.00000000e+000,
    E                 -1.00000000e+000],
    E                [-1.00000000e+000, -1.00000000e+000, -1.00000000e+000,
    E                 -1.00000000e+000],
    E                [-1.00000000e+000, -1.00000000e+000, -1.00000000e+000,
    E                 -1.00000000e+000],
    E                [-1.00000000e+000, -1.00000000e+000, -1.00000000e+000,
    E                 -1.00000000e+000],
    E                [-1.00000000e+000, -1.00000000e+000, -1.00000000e+000,
    E                 -1.00000000e+000],
    E                [-1.00000000e+000, -1.00000000e+000, -1.00000000e+000,
    E                 -1.00000000e+000],
    E                [-1.00000000e+000, -1.00000000e+000, -1.00000000e+000,
    E                 -2.44758[124](https://github.com/unifyai/ivy/actions/runs/3705644627/jobs/6279825331#step:7:125)e-308]])]),
    E       as_variable=[False],
    E       native_array=[True],
    E       num_positional_args=1,
    E       with_out=False,
    E       container_flags=[True],
    E       instance_method=False,
    E       backend_fw=<module 'ivy.functional.backends.tensorflow' from '/ivy/ivy/functional/backends/tensorflow/__init__.py'>,
    E       fn_name='matrix_exp',
    E       ground_truth_backend='tensorflow',
    E   )
    E   
    E   You can reproduce this example by temporarily adding @reproduce_failure('6.55.0', b'AXicY2RkEGdkYWIAAkYGCGAklt1ACpuRETubDHuxsnGZj4vNgqSXBSgu8sh08VMgDxebgUT1uPSOsqnOPmzAwIDEH3D3jLJpzQZm5f9XZ4LADKgwAGd3Itg=') as a decorator on your test case

File: ivy/container/experimental/linear_algebra.py [hunk 4]
```diff
@@ -171,6 +171,55 @@ def kron(
...
+    @staticmethod
+    def static_matrix_exp(
+        x: Union[ivy.Array, ivy.NativeArray, ivy.Container],
+        /,
+        *,
+        key_chains: Optional[Union[List[str], Dict[str, str]]] = None,
+        to_apply: bool = True,
+        out: Optional[ivy.Container] = None,
+    ) -> ivy.Container:
+        return ContainerBase.multi_map_in_static_method(
```
Comment (Daniel4078) [inline ivy/container/experimental/linear_algebra.py:183]: So the error happens at this line and it suggests that "multiple_map_in_static_method" does not exits for ContainerBase.

Comment (Daniel4078) [reply ivy/container/experimental/linear_algebra.py:183]: check the "static_kron" and "diagflat" just before this function where it uses "cont_multi_map_in_static_method" instead

Comment (anurag629) [reply ivy/container/experimental/linear_algebra.py]: Ohh! I missed that but now it's corrected.

**Review by Daniel4078 — CHANGES_REQUESTED** [review]
I found only one error left for you to fix:
numpy:    ivy.exceptions.IvyException: Cannot convert to numpy dtype. bfloat16 is not supported by NumPy.

**Comment by anurag629** [comment]
> I found only one error left for you to fix: numpy: ivy.exceptions.IvyException: Cannot convert to numpy dtype. bfloat16 is not supported by NumPy.

@Daniel4078 I added <code>@with_unsupported_dtypes({"1.23.0 and below": ("bfloat16",)}, backend_version)</code> decorator in numpy but it is showing same error. Why??

**Comment by Daniel4078** [comment]
> > I found only one error left for you to fix: numpy: ivy.exceptions.IvyException: Cannot convert to numpy dtype. bfloat16 is not supported by NumPy.
> 
> @Daniel4078 I added `@with_unsupported_dtypes({"1.23.0 and below": ("bfloat16",)}, backend_version)` decorator in numpy but it is showing same error. Why??

well, here is an alternative way: like what the eip function did you can add a type convertion before calling the actural function, which will even make bfloat16 supported for our numpy version.
  like  if ivy.dtype(x) == ivy.bfloat16:
        x = x.astype(np.float16)

**Review by Daniel4078 — APPROVED** [review]
LGTM

**Pull request merged**